### PR TITLE
fix: Parser - Equal key name replace conflict

### LINKED ISSUE #9245: Bug: Parser - Equal key name replace conflict
### PHP Version

8.3

### CodeIgniter4 Version

4.5.5

### CodeIgniter4 Installation Method

Composer (using `codeigniter4/appstarter`)

### Which operating systems have you tested for this bug?

macOS

### Which server did you use?

apache

### Database

MySQL 8.3.0

### What happened?

Using the same key name in an array although on different levels, leads to a premature substitution of `{key}`, when that `{key}` is referring in a nested one on a loop substitution.

(Based on the Steps to Reproduce section)

With the current Parser behaviour, both ```{id}``` vars will be replaced with their respective value but given that the second ```{id}``` is inside ```{blog_entries}```, by 'blindly' replacing all instances of ```{variable}```, ```{blog_entries}{id}{/blog_entries}``` will turn into ```{blog_entries}1{/blog_entries}``` which will fail to replace the ```{blog_entries}{id}{/blog_entries}``` with the proper data when it reaches to the blog_entries array.

### Steps to Reproduce

```
$data = [
    'id'   => 'id_1',
    'blog_title'   => 'My Blog Title',
    'blog_heading' => 'My Blog Heading',
    'blog_entries' => [
        ['id' => 'sub_id_1', 'title' => 'Title 1', 'body' => 'Body 1'],
        ['id' => 'sub_id_2', 'title' => 'Title 2', 'body' => 'Body 2'],
        ['id' => 'sub_id_3', 'title' => 'Title 3', 'body' => 'Body 3'],
        ['id' => 'sub_id_4', 'title' => 'Title 4', 'body' => 'Body 4'],
        ['id' => 'sub_id_5', 'title' => 'Title 5', 'body' => 'Body 5']
    ]
];

$template = <<<EOD
    ID: {id}
    title: {blog_title}
    Blog Entries:
    {blog_entries}
    {id} - {title} - {body}
{/blog_entries}
EOD;

$parser = service('parser');

print $parser->setData($data)->renderString($template);
```

Returns:

```
ID: id_1
title: My Blog Title Blog 
Entries: 
id_1 - Title 1 - Body 1
id_1 - Title 2 - Body 2
id_1 - Title 3 - Body 3
id_1 - Title 4 - Body 4
id_1 - Title 5 - Body 5
```

Going a level down

```
$data = [
    'id'   => 'id_1',
    'blog_title'   => 'My Blog Title',
    'blog_heading' => 'My Blog Heading',
    'blog_entries' => [
        ['items' =>
            [
                ['id' => 'sub_id_1', 'title' => 'Title 1', 'body' => 'Body 1'],
                ['id' => 'sub_id_2', 'title' => 'Title 2', 'body' => 'Body 2'],
                ['id' => 'sub_id_3', 'title' => 'Title 3', 'body' => 'Body 3'],
                ['id' => 'sub_id_4', 'title' => 'Title 4', 'body' => 'Body 4'],
                ['id' => 'sub_id_5', 'title' => 'Title 5', 'body' => 'Body 5']
            ]
        ]
    ]
];

$template = <<<EOD
  ID: {id}
  title: {blog_title}
  Blog Entries:
  {blog_entries}
{items}
  {id} - {title} - {body}
{/items}
{/blog_entries}
EOD;

$parser = service('parser');

echo $parser->setData($data)->renderString($template);
```

Returns:

```
ID: id_1
title: My Blog Title
Blog Entries: 
{items} id_1 - {title} - {body} {/items}
```

And it fails to properly replace the rest of the data.

### Expected Output

```
ID: id_1
title: My Blog Title
Blog Entries: 
sub_id_1 - Title 1 - Body 1
sub_id_2 - Title 2 - Body 2
sub_id_3 - Title 3 - Body 3
sub_id_4 - Title 4 - Body 4
sub_id_5 - Title 5 - Body 5
```

### Anything else?

This behaviour seems to stem from how the ```Parser::parse()``` function acts. From my understanding, when the parsing occurs the method loops over the data passed to it and if the data to be replaced isn't an array (Pair) then, it immediately replaces the content (Single). Which means, if there is multiple keys with the same identifier down the line, they will get replaced all at once, breaking further processing specially when using Pairs like ```{blog_entries}{items}{id}{items}{/blog_entries}```.

I propose a [solution](https://github.com/cosdiabos/CodeIgniter4/commit/2dbcac8725f9b03774d798508e6e8a0ed3476ed8) in which the return of ```parsePair``` and ```parseSingle``` are stored independently, then merged together (Pairs with Single), and finally the replacement loop happens. With this approach, the replacement happens from the Pairs to Single and the output is as expected since ```{blog_entries}{items}{id}{items}{/blog_entries}``` will be ```{blog_entries}{items}sub_id_1{items}{/blog_entries}``` before the top ```{id}``` replace is triggered.

### PR DESCRIPTION
**Pull request by CosDiabos**
**Description**
Fixes the conflict when replacing View Parser {variables} in the provided data with the same key name, even when on different levels.

This fix focus on ```parse()``` method, in which the return of ```parsePair``` and ```parseSingle``` methods are stored independently , then merged together (Pairs with Single), and finally replaced. With this approach, the replacement happens from the Pairs to Single and the output is as expected.

More details and examples in issue #9245 

Fixes #9245 

**Checklist:**
- [ ] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [ ] Conforms to style guide

### CODE REVIEW

**Comment by samsonasik** [comment]
Run Rector and cs fix:

```
vendor/bin/rector && composer cs-fix
```

also, a test case seems needed.

**Comment by samsonasik** [comment]
You can rename PR title with `fix: ` prefix

File: system/View/Parser.php [hunk 1]
```diff
@@ -254,20 +254,33 @@ protected function parse(string $template, array $data = [], ?array $options = n
...
+        // loop over the data variables, saving regex and data
+        // for later replacement.
         foreach ($data as $key => $val) {
             $escape = true;
 
             if (is_array($val)) {
-                $escape  = false;
-                $replace = $this->parsePair($key, $val, $template);
+                $escape              = false;
+                $replacePairsStack[] = ['replace' => $this->parsePair($key, $val, $template), 'escape' => $escape];
```
Comment (mostafakhudair) [inline system/View/Parser.php:268]:
```suggestion
                $escape              = false;
                $replacePairsStack[] = [
                    'replace' => $this->parsePair($key, $val, $template),
                    'escape'  => $escape,
                ];
```

File: system/View/Parser.php [hunk 2]
```diff
@@ -254,20 +254,33 @@ protected function parse(string $template, array $data = [], ?array $options = n
...
             } else {
-                $replace = $this->parseSingle($key, (string) $val);
+                $replaceSingleStack[] = ['replace' => $this->parseSingle($key, (string) $val), 'escape' => $escape];
             }
```
Comment (mostafakhudair) [inline system/View/Parser.php:271]:
```suggestion
                $replaceSingleStack[] = [
                    'replace' => $this->parseSingle($key, (string) $val),
                    'escape'  => $escape,
                ];
            }
```

**Comment by datamweb** [comment]
@CosDiabos To ensure the security and authenticity of your commits, please sign **all** of them with GPG. This shows others that the commits were genuinely made by you and can be trusted.

For a detailed guide on signing commits with GPG, please refer to [gpg-signing-old-commits](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#gpg-signing-old-commits).

**Comment by CosDiabos** [comment]
Thank you all for your help!

**Review by michalsn — CHANGES_REQUESTED** [review]
@CosDiabos Looks great! We need one last thing - a changelog entry.

Please add it under the Bugs Fixed section, here: https://github.com/codeigniter4/CodeIgniter4/blob/develop/user_guide_src/source/changelogs/v4.5.6.rst

Something like: 
> **Parser**: Fixed a bug that caused equal key names to be replaced by the key defined first.

Thanks.

**Review by michalsn — APPROVED** [review]

**Comment by michalsn** [comment]
@CosDiabos Thank you for your first contribution!

**Pull request merged**